### PR TITLE
bench: add benchmark for wallet 'AvailableCoins' function.

### PR DESCRIPTION
**Pull request by furszy**
#### Rationale

`AvailableCoins` is part of several important flows for the wallet; from RPC commands that create transactions like `fundrawtransaction`, `send`, `walletcreatefundedpsbt`, get the available balance, list the available coins with `listunspent` etc. to GUI connected processes that perform the same or similar actions: tx creation, available balance calculation, present the spendable coins in the coin control dialog.

As we are improving this process in #24699, #25005 and there are more structural changes coming on the way. This benchmark aims to ensure us that, at least, there are no regressions (obviously performance improvements are great but, at least for me, this heads into the direction of having a base metric to compare future structural changes).

#### Implementation Notes

There are 5 new benchmarks, one per wallet supported output type (LEGACY, P2SH_SEGWIT, BECH32, BECH32M), plus a multi-output-type wallet benchmark which contains outputs from all the descriptor types.

The test, by default, fills-up the wallet with 1k transactions, 2k outputs. Mainly to not consume much time if the user just want to verify that no substantial regressions were introduced. But, my expectation for those who are focused on this process is to use a much higher number locally to really note the differences across commits.

### CODE REVIEW

File: src/test/util/wallet.cpp [hunk 1]
```diff
@@ -20,11 +20,15 @@ const std::string ADDRESS_BCRT1_UNSPENDABLE = "bcrt1qqqqqqqqqqqqqqqqqqqqqqqqqqqq
 std::string getnewaddress(CWallet& w)
 {
     constexpr auto output_type = OutputType::BECH32;
+    return EncodeDestination(getnewaddress(w, output_type));
+}
+
+CTxDestination getnewaddress(CWallet& w, OutputType output_type)
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/test/util/wallet.cpp:26]: yeah agree

**Comment by furszy** [comment]
Thanks laanwj for the review!

It's interesting to see other results, locally this are my results:

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      132,289,908.40 |                7.56 |    0.3% |      7.27 | `WalletAvailableCoinsMulti`
|      104,691,758.40 |                9.55 |    0.1% |      5.77 | `WalletAvailableCoinsOnlyBech32`
|      196,688,875.00 |                5.08 |    0.2% |     10.82 | `WalletAvailableCoinsOnlyBech32M`
|       70,386,791.80 |               14.21 |    0.2% |      3.88 | `WalletAvailableCoinsOnlyLegacy`
|      111,477,558.20 |                8.97 |    0.1% |      6.14 | `WalletAvailableCoinsOnlyP2SH_SEGWIT`
```

But let me do something, going to decrease the bench number of iterations from 5 to 2. It should decrease the total time by half.

**Comment by furszy** [comment]
Done, decreased the number of iterations to 2. And this are the results:

```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      129,406,083.50 |                7.73 |    0.0% |      2.85 | `WalletAvailableCoinsMulti`
|      105,704,646.00 |                9.46 |    0.2% |      2.33 | `WalletAvailableCoinsOnlyBech32`
|      206,642,667.00 |                4.84 |    0.6% |      4.55 | `WalletAvailableCoinsOnlyBech32M`
|       69,838,895.50 |               14.32 |    0.0% |      1.54 | `WalletAvailableCoinsOnlyLegacy`
|      112,263,500.00 |                8.91 |    0.0% |      2.47 | `WalletAvailableCoinsOnlyP2SH_SEGWIT`
```

Total time: ~15 seconds to run the 5 new benchmarks at most here now.
Previously, with 5 iterations total time was ~40 seconds. So, you should see more or less a 40% total time decrease there as well.

**Comment by furszy** [comment]
Aside from that, I think that a good future question to ask ourselves would be whether we want to always run all the benchmarks all the time or not. Because, sooner or later, as the project isn't going to stop growing, we could end up sacrificing accuracy, implementing not entirely real scenarios (skipping parts of the flow that is being benchmarked) in favor of decreasing the total bench time (there is where my comment in https://github.com/bitcoin/bitcoin/pull/24924#pullrequestreview-985000006 heads to for example).

But well, this is just me thinking out loud.. after #24924, the total bench time will be pretty decent to not have to worry about this for a while.

File: src/bench/wallet_available_coins.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    CWallet wallet{test_setup->m_node.chain.get(), "", gArgs, CreateMockWalletDatabase()};
+    {
+        LOCK(wallet.cs_wallet);
+        if (std::any_of(output_type.begin(), output_type.end(), [](OutputType type){return type > OutputType::LEGACY;})) {
+            wallet.SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+            wallet.SetupDescriptorScriptPubKeyMans();
+        }
+        if (std::any_of(output_type.begin(), output_type.end(), [](OutputType type){return type == OutputType::LEGACY;})) {
+            wallet.SetupLegacyScriptPubKeyMan();
+        }
```
Comment (achow101) [inline src/bench/wallet_available_coins.cpp:86]: In b791c2ae2eb02ff53762ac5ed2b355e5248ac041 "bench: add benchmark for wallet 'AvailableCoins' function."

The type of `ScriptPubKeyMan` is not related to the `OutputType`, so having this depend on the OutputType does not make sense. Unfortunately the word "legacy" refers to a number of different things within the wallet.

The type of wallet is only going to have an effect on how `IsMine` performs, and I don't think that really matters for this measurement.

File: src/bench/wallet_available_coins.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+static void WalletAvailableCoinsOnlyBech32(benchmark::Bench& bench) { WalletAvailableCoins(bench, {OutputType::BECH32}); }
+static void WalletAvailableCoinsOnlyP2SH_SEGWIT(benchmark::Bench& bench) { WalletAvailableCoins(bench, {OutputType::P2SH_SEGWIT}); }
+static void WalletAvailableCoinsOnlyLegacy(benchmark::Bench& bench) { WalletAvailableCoins(bench, {OutputType::LEGACY}); }
+static void WalletAvailableCoinsMulti(benchmark::Bench& bench) { WalletAvailableCoins(bench, {OutputType::LEGACY, OutputType::BECH32M, OutputType::BECH32, OutputType::P2SH_SEGWIT}); }
+
+BENCHMARK(WalletAvailableCoinsOnlyBech32M);
+BENCHMARK(WalletAvailableCoinsOnlyBech32);
+BENCHMARK(WalletAvailableCoinsOnlyP2SH_SEGWIT);
+BENCHMARK(WalletAvailableCoinsOnlyLegacy);
+BENCHMARK(WalletAvailableCoinsMulti);
```
Comment (achow101) [inline src/bench/wallet_available_coins.cpp:128]: In b791c2ae2eb02ff53762ac5ed2b355e5248ac041 "bench: add benchmark for wallet 'AvailableCoins' function."

I don't really think it is useful to have `AvailableCoins` run for different `OutputTypes`. There really isn't any difference for how `AvailableCoins` behaves depending on the `OutputType`.

Comment (furszy) [reply src/bench/wallet_available_coins.cpp:128]: Hmm, wouldn't we be having the same benchmark times for all the cases if that would be the case? (we are adding the same amount of txs and outputs to each of them).

I mean, your #24699 improves exactly this. The time difference there is for the solving provider and `IsMine` related calls.

Comment (achow101) [reply src/bench/wallet_available_coins.cpp:127]: Yes, it should be the same benchmark times for all cases, so it doesn't make sense to run it separately for each `OutputType`.

**Review by hernanmarino — COMMENTED** [review]
Approach ACK fd583c8184ef2a270e7df13360f78fd9dda258a0. @furszy are you planning on rebasing this one, now that #25685 was merged ?

**Comment by pablomartin4btc** [comment]
> Aside from that, I think that a good future question to ask ourselves would be whether we want to always run all the benchmarks all the time or not. Because, sooner or later, as the project isn't going to stop growing, we could end up sacrificing accuracy, implementing not entirely real scenarios...

I can see you also introduced that idea and the improvement is already merged and available; [good stuff](https://github.com/bitcoin/bitcoin/pull/26158)!

**Review by pablomartin4btc — COMMENTED** [review]
Code review ACK.
I've verified most changes were already merged on #25685, waiting for your next commit to re ACK this PR again.

**Comment by furszy** [comment]
yeah, will rebase + rearrange it in the coming days. Thanks for the ping 👍🏼 .

File: src/bench/wallet_create_tx.cpp [hunk 4]
```diff
@@ -132,11 +132,52 @@ static void WalletCreateTx(benchmark::Bench& bench, const OutputType output_type
...
+    }
+
+    // Check available balance
+    auto bal = wallet::GetAvailableBalance(wallet); // Cache
+    assert(bal == 50 * COIN * (chain_size - COINBASE_MATURITY));
+
+    bench.epochIterations(2).run([&] {
+        LOCK(wallet.cs_wallet);
+        const auto& res = wallet::AvailableCoins(wallet);
+        std::cout << res.All().size() << std::endl;
```
Comment (achow101) [inline src/bench/wallet_create_tx.cpp:168]: I think this is a debugging leftover?

Comment (furszy) [reply src/bench/wallet_create_tx.cpp]: it's all about perspective, could be a feature too.

fixed, thanks.

**Comment by achow101** [comment]
ACK 3a4f8bc24271d05765e9bf1e26558a28ab2e6b81

**Review by hernanmarino — COMMENTED** [review]
ACK 3a4f8bc24271d05765e9bf1e26558a28ab2e6b81

**Review by hernanmarino — APPROVED** [review]

**Comment by hernanmarino** [comment]
This is my output : 

```
|               ns/op |                op/s |    err% |          ins/op |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|---------------:|--------:|----------:|:----------
|       10,914,205.50 |               91.62 |    0.1% |  136,775,849.50 |  12,819,423.00 |    0.0% |      0.24 | `WalletAvailableCoins`

```

**Review by pablomartin4btc — COMMENTED** [review]
tested ACK.

Here my results:

```|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       20,730,690.00 |               48.24 |    3.8% |      0.47 | `WalletAvailableCoins`
```

From time to time I get recommendations to increase the iters as it warns unstability:

```|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       20,150,356.50 |               49.63 |    5.3% |      0.48 | :wavy_dash: `WalletAvailableCoins` (Unstable with ~2.0 iters. Increase `minEpochIterations` to e.g. 20)
```

Not sure if this could be an issue.

**Review by aureleoules — APPROVED** [review]
ACK 3a4f8bc24271d05765e9bf1e26558a28ab2e6b81

**Pull request merged**